docs/refman/ppr-config.8.pod
#
# mouse:~ppr/src/docs/refman/ppr-config.8.pod
# Copyright 1995--2005, Trinity College Computing Center.
# Written by David Chappell.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.
# 
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
#
# Last modified 17 January 2005.
#

=head1 NAME

B<ppr-config> - Configuration tool for PPR addons

=head1 SYNOPSIS

B<ppr-config> [--version] [--mk] [--sh] [--ph] [--pm] [--confdir] [--libdir] [---sharedir] [--var-spool-ppr]

=head1 DESCRIPTION

This program will print configuration information about the installed copy
of PPR.  Programs that extend PPR should use it to locate PPR's installation
directories and to obtain other configuration information.

=over 4

=item --version

Print the PPR version number.

=item --mk

Print a makefile fragment, suitable for including in a makefile.

=item --sh

Print a shell script fragment, suitable for including in a Borne shell script.

=item --ph

Print a Perl header file containing PPR configuration information, suitable
for including using B<require>.  (PPR's F<paths.ph> is created by this
command.)

=item --pm

Print a Perl module containing PPR configuration information, suitable for
including using B<use>.  (PPR's F<paths.pm> is created by this command.)

=item --tcl

Print a Tcl script fragment, suitable for including in a Tcl script.

=item --confdir

Print the path of PPR's configuration directory.

=item --libdir

Print the path of PPR's library directory.

=item --sharedir

Print the path of PPR's share directory.

=item --var-spool-ppr

Print the path of PPR's spool directory.

=back

=head1 SEE ALSO

"PPR, a PostScript Print Spooler"

=head1 HISTORY

PPR was written at Trinity College during 1993--2003.

=head1 AUTHOR

David Chappell, Trinity College Computing Center, Hartford, Connecticut.

=cut


History.txt
mouse:~ppr/src/History.txt (symbolic link)

This file contains a history of changes to PPR which occured after the
release of version 1.00.


1.00, 26 April 1995

First public release.


1.10, 27 April 1995--7 June 1995

Removed "From:" line from message passed to Sendmail by pprd when sending
a message listing the remaining printer faults.  This is so that Sendmail
will take the "real name" from the password file and put it in the "From:"
line.

Removed dotmatrix/input.epson, a GIF file which had been inadvertently
included in the distribution.

Moved the contents of include/atalk_streams.h into
interfaces/atalk_streams.c since it was not long enough to justify an
include file, nor was it of interest to any other program.

Added code to ppad to classify members of the group "ppradmin" as PPR
addministrators who can edit printer and group configurations.

Added -v switch to ppad.

Removed uid's 1-99 from the list of priveledged users.

Fixed ppad(8) so that it considers the group "ppradmin" as priveledged
rather than "ppr".

Changed ppr(1) so that it will accept the members of the group "pprprox" as
priveledged as well as the members of "ppr".  This is for security reasons
as one may not want to grant a user who has the privedge of using the -f
switch the priveledge of deleting other people's jobs.

Modified ppop(1) and pprd(8) so that user's can request to see the queue
entry for a single job.  Previously, the only options were all jobs for a
certain destination or all jobs in all queues.

Rewrote and renamed the undocumented "ppop qdebug" command to become the new
documented command "ppop details".  (This was done primarily to support
wwwppop, an experimental managment interface for PPR.  Since it was still
incomplete it was not released with this version of PPR.)

Renamed interfaces/atalk_streams.c to atalk_ali.c and
papsrv/papsrv_streams.c to papsrv_ali.c

Removed code from "make dirs" so that it doesn't create the directory
"/etc/ppr/groups.ppd".

Added sleep=N and eoftimeout=N options to tcpip interface.

Added -C _filename_ switch to ppr invokation in xppr.

Added "-default okay" to xmessage invokation in xppr.

Added the "ppop lpq" sub-command to the "ppop help" information, thus
making it a documented sub-command.

Extended automatic bin selection to multiple media types when "%%PageMedia:"
comments are present.  This has not yet been well tested.

Made "%%EndProcset" an acceptable substitute for "%%EndProcSet" due to a
reputed bug in QuickDraw GX.

Changed rule in ppr_infile.c for detecting PCL code so that documents with
only one escape would still be detected.


1.20, 8 June 1995--10 January 1996

Fixed code in ppr_infile.c so that the ppr -o switch will work even if there
is no "DefFiltOpts:" line in the printer or group configuration file.

Added '(' to list of likely second characters for HP escape codes
in ppr_infile.c

Replaced printing of errno value with strerror(errno) output in some places,
eliminated printing of errno=x from other places where both were printed.

Fixed ppr(1) so that it sets the environment variables "PATH", "IFS", and
"SHELL" so that filters and responders it invokes are less easy to subvert.

Removed setting of PATH from ~ppr/src/misc_filters/tex.perl and changed
shell name from "/usr/bin/tainperl" to "/usr/bin/perl".

Changed code which determins default responder address so that the
environment variables "USER" and "LOGNAME" are used in preference to the
current uid.

Added prompts in Configure to set the directory into which the links to PPR
binaries will be written and the path to the X-Windows binaries such as
xterm.

Modified papsrv so that it will pass the network and node number of the
client to ppr with the -r switch.  (This is for use with an experimental
AppleTalk responder.)

Modified papsrv.c, renaming cleanup() to termination_handler() and causing
it to call fatal() only the first time it is called.  Without this, name
could be left on the network if the CAP version received two terminate
signals in rapid sucession.

Added code to re-line-wrap responder messages for certain responders.
(The library function responder_width() returns the width for a responder.
Currently it is hard coded to return zero for all responders except "samba".
A return value of zero indicates that the message should not be
re-line-wrapped.)

Dropt reading of LPDEST and PRINTER by ppr(1).  (I can not really remember
why I put this feature in.  I don't think the reasons where very good.  It
may have been that I thought that the same printer would be available thru
two spooling systems so one might want to set the default destination for
both with one variable.  That now seems a little far fetched.)

Separated Solaris and Starserver sections of sysdep.h

Added PNG (Portable Network Graphics) format with magic number recognition
to ppr_infile.c.  This remains untested since I do not have any programs
which manipulate this graphics format.

In pprd.c, put sleep(1) between execl("pprdrv") and _exit(242) so that pprd
has time to record the PID before the child exits.  (If this is not done
and execl() fails immediately, a child which should have become pprdrv may
exit before the parent records its PID in the information structure for that
queue.  This result is that it will not recoginize the purpose of the child
when it receives notification of its death and will leave the queue in the
printing state indefinitely.)

Did same in pprdrv.c in code which executes interface.  Also changed
_exit(255) to _exit(242).  (Exit 242 is the new standard since PERL uses 255
as the exit code when die() is executed.)

Added automatic creation of "DefFiltOpts:" lines to ppad(8).  Manual
creation is supported by the the "ppad deffiltopts" and "ppad group
deffiltopts" command.

Modified ppop so that interactive mode restores the terminal settings when
it exits.  Also changed terminal setting code to use the POSIX method of

makeprogs/ppr-config.c
/*
** mouse:~ppr/src/makeprogs/ppr-config.c
** Copyright 1995--2003, Trinity College Computing Center.
** Written by David Chappell.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**
** * Redistributions of source code must retain the above copyright notice,
** this list of conditions and the following disclaimer.
** 
** * Redistributions in binary form must reproduce the above copyright
** notice, this list of conditions and the following disclaimer in the
** documentation and/or other materials provided with the distribution.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
** AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
** IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
** ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
** LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
** CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
** SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
** CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
** ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
** POSSIBILITY OF SUCH DAMAGE.
**
** Last modified 7 March 2003.
*/

#include "before_system.h"
#include <stdio.h>
#include <string.h>
#include "gu.h"
#include "global_defines.h"
#include "global_structs.h"
#include "version.h"

const char myname[] = "ppr-config";

static void do_paths(char format[], char language[])
	{
	printf("# This is NOT a configuration file!	 Don't modify it!\n");

	if(strcmp(language, "--pm") == 0)
		printf("package PPR;\n");

	printf(format, "SHORT_VERSION", SHORT_VERSION);

	/*
	** These appear in all but paths.mk.  They don't appear in
	** paths.mk because they are already in global.mk.
	*/
	if(strcmp(language, "--mk") != 0)
		{
		printf(format, "USER_PPR", USER_PPR);
		printf(format, "USER_PPRWWW", USER_PPRWWW);
		printf(format, "GROUP_PPR", GROUP_PPR);
		printf(format, "CONFDIR", CONFDIR);
		printf(format, "HOMEDIR", HOMEDIR);
		printf(format, "SHAREDIR", SHAREDIR);
		printf(format, "VAR_SPOOL_PPR", VAR_SPOOL_PPR);
		printf(format, "TEMPDIR", TEMPDIR);
		printf(format, "SYSBINDIR", SYSBINDIR);

		/* non-path stuff */
		printf(format, "VERSION", VERSION);
		printf(format, "SHORT_VERSION", SHORT_VERSION);
		}

	/* Second Level Directories */
	printf(format, "PRCONF", PRCONF);
	printf(format, "GRCONF", GRCONF);
	printf(format, "ALIASCONF", ALIASCONF);
	printf(format, "ACLDIR", ACLDIR);
	printf(format, "LOGDIR", LOGDIR);
	printf(format, "RUNDIR", RUNDIR);
	printf(format, "INTDIR", INTDIR);
	printf(format, "PPDDIR", PPDDIR);
	printf(format, "RESPONDERDIR", RESPONDERDIR);

	/* Configuration Files */
	printf(format, "PPR_CONF", PPR_CONF);
	printf(format, "UPRINTCONF", UPRINTCONF);
	printf(format, "UPRINTREMOTECONF", UPRINTREMOTECONF);
	printf(format, "MEDIAFILE", MEDIAFILE);

	/* Search Path */
	printf(format, "SAFE_PATH", SAFE_PATH);

	/* Program Paths */
	printf(format, "PPR_PATH", PPR_PATH);
	printf(format, "PPOP_PATH", PPOP_PATH);
	printf(format, "PPAD_PATH", PPAD_PATH);
	printf(format, "PPR2SAMBA_PATH", PPR2SAMBA_PATH);
	printf(format, "TAIL_STATUS_PATH", TAIL_STATUS_PATH);
	/* printf(format, "PPJOB_PATH", PPJOB_PATH); */
	printf(format, "SENDMAIL_PATH", SENDMAIL_PATH);

	/* Internationalization */
	printf(format, "LOCALEDIR", LOCALEDIR);
	printf(format, "PACKAGE", PACKAGE);
	printf(format, "PACKAGE_INTERFACES", PACKAGE_INTERFACES);
	printf(format, "PACKAGE_PPRD", PACKAGE_PPRD);
	printf(format, "PACKAGE_PPRDRV", PACKAGE_PPRDRV);
	printf(format, "PACKAGE_PAPSRV", PACKAGE_PAPSRV);
	printf(format, "PACKAGE_PPRWWW", PACKAGE_PPRWWW);

	/* More stuff for PPR.pm. */
	if(strcmp(language, "--pm") == 0)
		{
		printf("$SIZEOF_struct_Media=%d;\n", (int)sizeof(struct Media));
		}

	/* Perl modules and libraries must end with something that evaluates to true. */
	if(strcmp(language, "--ph") == 0 || strcmp(language, "--pm") == 0)
		printf("1;\n");
	}

int main(int argc, char *argv[])
	{
	int i;

	if(argc < 2)
		{
		fprintf(stderr,
				"Usage: %s [OPTIONS]\n"
				"Options:\n"
				"\t--version\n"
				"\t--mk\n"
				"\t--sh\n"
				"\t--ph\n"
				"\t--pm\n"
				"\t--tcl\n"
				"\t--confdir\n"
				"\t--homedir\n"
				"\t--sharedir\n"
				"\t--var-spool-ppr\n",
				myname);
		return 1;
		}

	for(i=1; i<argc; i++)
		{
		if(strcmp(argv[i], "--version") == 0)
			printf("%s\n", SHORT_VERSION);
		else if(strcmp(argv[i], "--mk") == 0)
			do_paths("%s=%s\n", argv[i]);
		else if(strcmp(argv[i], "--sh") == 0)
			do_paths("%s=\"%s\"\n", argv[i]);
		else if(strcmp(argv[i], "--ph") == 0 || strcmp(argv[i], "--pm") == 0)
			do_paths("$%s=\"%s\";\n", argv[i]);
		else if(strcmp(argv[i], "--tcl") == 0)
			do_paths("set %s \"%s\"\n", argv[i]);
		else if(strcmp(argv[i], "--confdir") == 0)
			printf("%s\n", CONFDIR);
		else if(strcmp(argv[i], "--homedir") == 0)
			printf("%s\n", HOMEDIR);
		else if(strcmp(argv[i], "--sharedir") == 0)
			printf("%s\n", SHAREDIR);
		else if(strcmp(argv[i], "--var-spool-ppr") == 0)
			printf("%s\n", VAR_SPOOL_PPR);
		else if(strcmp(argv[i], "--tempdir") == 0)
			printf("%s\n", TEMPDIR);
		else
			{
			fprintf(stderr, "%s: unknown option %s\n", myname, argv[i]);
			return 1;
			}
		}

	return 0;
	}

/* end of file */
